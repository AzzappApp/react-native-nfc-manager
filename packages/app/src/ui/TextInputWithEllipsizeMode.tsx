import { useRef, useState } from 'react';
import { View } from 'react-native';
import { colors } from '#theme';
import { createStyleSheet, useStyleSheet } from '#helpers/createStyles';
import Text from './Text';
import { styleSheetData as textInputStyleSheet } from './TextInput';
import TextInputWithPrefix from './TextInputWithPrefix';
import type {
  TextInputProps,
  NativeSyntheticEvent,
  TextInputFocusEventData,
  ViewStyle,
  TextInput,
} from 'react-native';

export type TextInputWithEllipsizeModeProps = Pick<
  TextInputProps,
  | 'autoCapitalize'
  | 'enterKeyHint'
  | 'keyboardType'
  | 'onBlur'
  | 'onChangeText'
  | 'onFocus'
  | 'placeholder'
  | 'style'
  | 'value'
> & {
  /**
   * Whether the input is in error state
   */
  isErrored?: boolean;
  prefix?: string;
};

/**
 * A wrapper around TextInput that adds Azzapp's default styling.
 * This version is a workaround for missing ellipsizemode support on
 * TextInput.
 * It simplify swap between TextInput and Text (with ellipsize) when it is unfocused
 */
const TextInputWithEllipsizeMode = ({
  isErrored,
  onFocus,
  onBlur,
  style,
  prefix,
  ...props
}: TextInputWithEllipsizeModeProps) => {
  //#region hooks
  const styles = useStyleSheet(styleSheet);
  const [isFocused, setIsFocused] = useState(false);
  //#endregion

  const nativeTextInputRef = useRef<TextInput>(null);

  const onFocusInner = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setIsFocused(true);
    onFocus?.(e);
  };

  const onBlurInner = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setIsFocused(false);
    onBlur?.(e);
  };

  const onTextPress = () => {
    nativeTextInputRef?.current?.focus();
  };

  return (
    <View style={style as ViewStyle}>
      <TextInputWithPrefix
        testID="nativeInputText"
        ref={nativeTextInputRef}
        selectionColor={colors.primary400}
        {...props}
        placeholderTextColor={colors.grey400}
        onFocus={onFocusInner}
        onBlur={onBlurInner}
        allowFontScaling={false}
        style={styles.input}
        inputStyle={[
          style,
          isFocused && styles.focused,
          isErrored && styles.errored,
          !isFocused ? { color: 'transparent' } : undefined,
        ]}
        prefix={prefix}
      />
      {!isFocused ? (
        <Text
          onPress={onTextPress}
          {...props}
          allowFontScaling={false}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={[
            styles.input,
            style,
            isErrored && styles.errored,
            styles.inputText,
            {
              opacity: !isFocused ? 1 : 0,
              borderWidth: 0,
            },
          ]}
        >
          {props.value}
        </Text>
      ) : undefined}
    </View>
  );
};

const styleSheet = createStyleSheet(appearance => ({
  ...textInputStyleSheet(appearance),
  inputText: {
    top: 9, // workaround for text positioning from react native
    backgroundColor: 'transparent',
    justifyContent: 'center',
    position: 'absolute',
  },
}));

export default TextInputWithEllipsizeMode;
