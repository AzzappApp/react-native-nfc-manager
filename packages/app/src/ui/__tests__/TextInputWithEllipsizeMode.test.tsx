import { fireEvent, render, screen } from '@testing-library/react-native';
import TextInputWithEllipsizeMode from '../TextInputWithEllipsizeMode';

const mockedUseColorScheme = jest.fn();

jest.mock('react-native/Libraries/Utilities/useColorScheme', () => {
  return {
    default: mockedUseColorScheme,
  };
});

describe('TextInputWithEllipsizeMode', () => {
  test('empty component', () => {
    render(<TextInputWithEllipsizeMode />);
    expect(screen).toMatchSnapshot();
  });
  test('component with value', () => {
    render(<TextInputWithEllipsizeMode value="test" />);
    expect(screen).toMatchSnapshot();
  });
  test('component change text value', () => {
    const component = render(<TextInputWithEllipsizeMode value="test" />);
    const inputText = component.getByTestId('nativeInputText');
    fireEvent.changeText(inputText, 'text');
    expect(screen).toMatchSnapshot();
  });
});
