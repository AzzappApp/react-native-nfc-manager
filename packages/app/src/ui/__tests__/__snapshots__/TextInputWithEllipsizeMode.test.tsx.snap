// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputWithEllipsizeMode component change text value 1`] = `
<View>
  <View
    style={
      [
        {
          "backgroundColor": "#F5F5F6",
          "borderColor": "#F5F5F6",
          "borderRadius": 12,
          "borderWidth": 1,
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
          "fontSize": 16,
          "height": 47,
          "lineHeight": 20,
          "paddingHorizontal": 15,
          "paddingVertical": 3,
        },
        false,
        undefined,
        {
          "alignItems": "center",
          "flexDirection": "row",
        },
        {
          "backgroundColor": "#F5F5F6",
          "borderColor": "#F5F5F6",
          "borderRadius": 12,
          "borderWidth": 1,
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
          "fontSize": 16,
          "height": 47,
          "lineHeight": 20,
          "paddingHorizontal": 15,
          "paddingVertical": 3,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={false}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#A1A1A5"
      selectionColor="#FF502E"
      style={
        [
          {
            "color": "#0E1216",
            "flex": 1,
          },
          [
            undefined,
            false,
            undefined,
            {
              "color": "transparent",
            },
          ],
        ]
      }
      testID="nativeInputText"
      value="test"
    />
  </View>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    maxFontSizeMultiplier={1}
    numberOfLines={1}
    onPress={[Function]}
    style={
      [
        {
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
        },
        [
          {
            "backgroundColor": "#F5F5F6",
            "borderColor": "#F5F5F6",
            "borderRadius": 12,
            "borderWidth": 1,
            "color": "#0E1216",
            "fontFamily": "PlusJakartaSans-Regular",
            "fontSize": 16,
            "height": 47,
            "lineHeight": 20,
            "paddingHorizontal": 15,
            "paddingVertical": 3,
          },
          undefined,
          undefined,
          {
            "backgroundColor": "transparent",
            "justifyContent": "center",
            "position": "absolute",
            "top": 9,
          },
          {
            "borderWidth": 0,
            "opacity": 1,
          },
        ],
      ]
    }
    value="test"
  >
    test
  </Text>
</View>
`;

exports[`TextInputWithEllipsizeMode component with value 1`] = `
<View>
  <View
    style={
      [
        {
          "backgroundColor": "#F5F5F6",
          "borderColor": "#F5F5F6",
          "borderRadius": 12,
          "borderWidth": 1,
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
          "fontSize": 16,
          "height": 47,
          "lineHeight": 20,
          "paddingHorizontal": 15,
          "paddingVertical": 3,
        },
        false,
        undefined,
        {
          "alignItems": "center",
          "flexDirection": "row",
        },
        {
          "backgroundColor": "#F5F5F6",
          "borderColor": "#F5F5F6",
          "borderRadius": 12,
          "borderWidth": 1,
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
          "fontSize": 16,
          "height": 47,
          "lineHeight": 20,
          "paddingHorizontal": 15,
          "paddingVertical": 3,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={false}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#A1A1A5"
      selectionColor="#FF502E"
      style={
        [
          {
            "color": "#0E1216",
            "flex": 1,
          },
          [
            undefined,
            false,
            undefined,
            {
              "color": "transparent",
            },
          ],
        ]
      }
      testID="nativeInputText"
      value="test"
    />
  </View>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    maxFontSizeMultiplier={1}
    numberOfLines={1}
    onPress={[Function]}
    style={
      [
        {
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
        },
        [
          {
            "backgroundColor": "#F5F5F6",
            "borderColor": "#F5F5F6",
            "borderRadius": 12,
            "borderWidth": 1,
            "color": "#0E1216",
            "fontFamily": "PlusJakartaSans-Regular",
            "fontSize": 16,
            "height": 47,
            "lineHeight": 20,
            "paddingHorizontal": 15,
            "paddingVertical": 3,
          },
          undefined,
          undefined,
          {
            "backgroundColor": "transparent",
            "justifyContent": "center",
            "position": "absolute",
            "top": 9,
          },
          {
            "borderWidth": 0,
            "opacity": 1,
          },
        ],
      ]
    }
    value="test"
  >
    test
  </Text>
</View>
`;

exports[`TextInputWithEllipsizeMode empty component 1`] = `
<View>
  <View
    style={
      [
        {
          "backgroundColor": "#F5F5F6",
          "borderColor": "#F5F5F6",
          "borderRadius": 12,
          "borderWidth": 1,
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
          "fontSize": 16,
          "height": 47,
          "lineHeight": 20,
          "paddingHorizontal": 15,
          "paddingVertical": 3,
        },
        false,
        undefined,
        {
          "alignItems": "center",
          "flexDirection": "row",
        },
        {
          "backgroundColor": "#F5F5F6",
          "borderColor": "#F5F5F6",
          "borderRadius": 12,
          "borderWidth": 1,
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
          "fontSize": 16,
          "height": 47,
          "lineHeight": 20,
          "paddingHorizontal": 15,
          "paddingVertical": 3,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={false}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#A1A1A5"
      selectionColor="#FF502E"
      style={
        [
          {
            "color": "#0E1216",
            "flex": 1,
          },
          [
            undefined,
            false,
            undefined,
            {
              "color": "transparent",
            },
          ],
        ]
      }
      testID="nativeInputText"
    />
  </View>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    maxFontSizeMultiplier={1}
    numberOfLines={1}
    onPress={[Function]}
    style={
      [
        {
          "color": "#0E1216",
          "fontFamily": "PlusJakartaSans-Regular",
        },
        [
          {
            "backgroundColor": "#F5F5F6",
            "borderColor": "#F5F5F6",
            "borderRadius": 12,
            "borderWidth": 1,
            "color": "#0E1216",
            "fontFamily": "PlusJakartaSans-Regular",
            "fontSize": 16,
            "height": 47,
            "lineHeight": 20,
            "paddingHorizontal": 15,
            "paddingVertical": 3,
          },
          undefined,
          undefined,
          {
            "backgroundColor": "transparent",
            "justifyContent": "center",
            "position": "absolute",
            "top": 9,
          },
          {
            "borderWidth": 0,
            "opacity": 1,
          },
        ],
      ]
    }
  />
</View>
`;
